Tests/PostgresNIOTests/New/PostgresConnectionTests.swift

```diff
@@ -638,6 +638,63 @@ class PostgresConnectionTests: XCTestCase {
         }
     }
 
+    func testQueriesFailIfConnectionIsClosed() async throws {
+        let (connection, channel) = try await self.makeTestConnectionWithAsyncTestingChannel()
+
+        try await connection.closeGracefully()
+
+        XCTAssertEqual(channel.isActive, false)
+
+        do {
+            _ = try await connection.query("SELECT version;", logger: self.logger)
+            XCTFail("Expected to fail")
+        } catch let error as ChannelError {
+            XCTAssertEqual(error, .ioOnClosedChannel)
+        }
+    }
+
+    func testListenFailsIfConnectionIsClosed() async throws {
+        let (connection, channel) = try await self.makeTestConnectionWithAsyncTestingChannel()
+
+        try await connection.closeGracefully()
+
+        XCTAssertEqual(channel.isActive, false)
+
+        do {
+            _ = try await connection.listen("test_channel")
+            XCTFail("Expected to fail")
+        } catch let error as ChannelError {
+            XCTAssertEqual(error, .ioOnClosedChannel)
+        }
+    }
+
+    func testListenFailsIfConnectionIsClosedMidway() async throws {
+        let (connection, channel) = try await self.makeTestConnectionWithAsyncTestingChannel()
+
+        try await withThrowingTaskGroup(of: Void.self) { taskGroup in
+            taskGroup.addTask {
+                do {
+                    let listenSequence = try await connection.listen("test_channel")
+
+                    for try await _ in listenSequence {
+                        XCTFail("Expected to fail")
+                    }
+                } catch let error as PSQLError {
+                    XCTAssertEqual(error.code, .listenFailed)
+                }
+            }
+
+            taskGroup.addTask {
+                try await Task.sleep(for: .seconds(3))
+
+                try await connection.close().get()
```

Comment:
Using `.close().get()` instead of `.closeGracefully()` here because `.closeGracefully()` will wait it out till infinitely and beyond since it's "graceful" and the listener is still there.